resource/reapi/bindings/go/src/fluxcli/reapi_cli.go

```diff
@@ -14,226 +14,260 @@ package fluxcli
 #include "reapi_cli.h"
 */
 import "C"
+import (
+	"fmt"
+	"unsafe"
+)
 
 type (
 	ReapiCtx C.struct_reapi_cli_ctx_t
 )
 
-/*! Create and initialize reapi_cli context
-reapi_cli_ctx_t *reapi_cli_new ();
-*/
-
+// NewReapiCli creates a new resource API client
+// reapi_cli_ctx_t *reapi_cli_new ();
 func NewReapiCli() *ReapiCtx {
 	return (*ReapiCtx)(C.reapi_cli_new())
 }
 
+// // ReapiCliGetNode retruns a resource node
+// This function is currently not being used
 // func ReapiCliGetNode(ctx *ReapiCtx) string {
 // 	return C.GoString(C.reapi_cli_get_node((*C.struct_reapi_cli_ctx)(ctx)))
 // }
 
-/*! Destroy reapi cli context
-*
-* \param ctx           reapi_cli_ctx_t context object
-void reapi_cli_destroy (reapi_cli_ctx_t *ctx);
-*/
+// Given an integer return code, convert to go error
+// Also provide a meaningful string to the developer user
+func retvalToError(code int, message string) error {
+	if code == 0 {
+		return nil
+	}
+	return fmt.Errorf(message+" %d", code)
+}
 
+// ReapiCliDestroy destroys a resource API context
+// void reapi_cli_destroy (reapi_cli_ctx_t *ctx);
 func ReapiCliDestroy(ctx *ReapiCtx) {
 	C.reapi_cli_destroy((*C.struct_reapi_cli_ctx)(ctx))
 }
 
-/* int reapi_cli_initialize (reapi_cli_ctx_t *ctx, const char *jgf); */
+// ReapiCliInit initializes a new resource API context
+// int reapi_cli_initialize (reapi_cli_ctx_t *ctx, const char *jgf);
+func ReapiCliInit(ctx *ReapiCtx, jgf string, options string) (err error) {
 
-func ReapiCliInit(ctx *ReapiCtx, jgf string, options string) (err int) {
-	err = (int)(C.reapi_cli_initialize((*C.struct_reapi_cli_ctx)(ctx),
-		C.CString(jgf), (C.CString(options))))
-	return err
-}
+	jobgraph := C.CString(jgf)
+	defer C.free(unsafe.Pointer(jobgraph))
```

Comment:
I'm not sure that I'm correct here - e.g., https://github.com/golang/go/issues/43363. Going to call on @trws for feedback on this one. I think if we use CString without new it's generated on the stack and automatically freed ([reading here](https://stackoverflow.com/questions/10937421/need-to-delete-cstring-after-use-to-free-memory)) and I'm also seeing a pattern of wrapping defer in a function [reading here](https://groups.google.com/g/golang-nuts/c/9gKpG3nVblA). I am typically pattern matching and saw this defer pattern used a lot with cgo and assumed it was the right way to do things.

For a general defer, I always think of it as an invisible stack of pancakes that, whenever I defer something, I slap one new pancake on the top. And then there is a gopher somewhere that eats them, from the top down (so last in, first out) when the function returns. :pancakes: So I think for cases of CString without new, it's safe to remove the defer, and we can think about the others.

Reply:
That's a good way to think about it actually.  It's literally a dynamic stack of cleanup functors to run at the end of the function.  That's also the important difference from destructors, it's the end _of the function_ not the scope, and it's _dynamic_ so if you defer something in a loop, they will all stack up and run when the function exits.  I wrote up this little toy version for C when we were working on the WG14 proposal for defer if you want to poke at it.  It's more limited than the go version but has the same semantics: https://github.com/trws/libdefer

Reply:
@trws you make the coolest stuff!

I'm glad my pancake analogy didn't give you the crepes! :pancakes: 

I wound up removing the defers here because I read that for CString without new, it's put on the stack and thus cleaned up on function return. Also, we are going to work from https://github.com/flux-framework/flux-sched/pull/1062 (green!) because there were some testing issues here without the rebase.

Reply:
Thanks for the explanations, and that's awesome @trws wrote a library to prototype `defer` in C!